File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        result=[]
        if len(digits)==0:
            return result
        maps={}
        maps[0]='0'; maps[1]='1'; maps[2]='abc'; maps[3]='def'; maps[4]='ghi'; maps[5]='jkl'; maps[6]='mno';
        maps[7]='pqrs'; maps[8]='tuv'; maps[9]='wxyz';
        self.find_combinations(result,maps,digits,'',0)
        return result
    def find_combinations(self,result,maps,digits,current,index):
        if index==len(digits):
            result.append(current)
            return
        letters=maps[(int)(digits[index])]
        for i in range(len(letters)):
            self.find_combinations(result,maps,digits,current+letters[i],index+1)
        

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> result; 
        if(digits.size()==0) return result; 
        vector<string> mapping={"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
        findCombinations(result,digits,mapping,0,"");
        return result;
    }
    void findCombinations(vector<string> &result,string digits,vector<string>& mapping,int index,string current){
        if(index==digits.size()){
            result.push_back(current);
            return;
        }
        string letters=mapping[digits[index]-'0'];
        for(int i=0;i<letters.size();i++){
            findCombinations(result,digits,mapping,index+1,current+letters[i]);   
        }
    }
};